src/Database/LSMTree/Internal/Serialise/RawBytes.hs

```diff
@@ -214,16 +213,15 @@ fromByteString :: BS.ByteString -> RawBytes
 fromByteString = fromShortByteString . SBS.toShort
 
 -- | \( O(1) \) conversion from a strict bytestring to raw bytes.
+--
+-- Strict bytestrings are allocated using 'mallocPlainForeignPtrBytes', so we
+-- are expecting a 'PlainPtr' (or 'FinalPtr' with length 0).
+-- For other variants, this function will fail.
 unsafeFromByteString :: HasCallStack => BS.ByteString -> RawBytes
-unsafeFromByteString (BS.Internal.BS (GHC.ForeignPtr _ contents) n) =
-    case contents of
-      -- Strict bytestrings are allocated using 'mallocPlainForeignPtrBytes', so
-      -- we are expecting a 'PlainPtr' here.
-      PlainPtr mba# -> case unsafeFreezeByteArray# mba# realWorld# of
-                   (# _, ba# #) -> RawBytes (P.Vector 0 n (ByteArray ba#))
-      FinalPtr {}        -> error ("unsafeFromByteString: expected plain pointer, got FinalPtr (length " Prelude.++ show n Prelude.++ ")")
-      MallocPtr {}       -> error ("unsafeFromByteString: expected plain pointer, got MallocPtr (length " Prelude.++ show n Prelude.++ ")")
-      PlainForeignPtr {} -> error ("unsafeFromByteString: expected plain pointer, got PlainForeignPtr (length " Prelude.++ show n Prelude.++ ")")
+unsafeFromByteString bs =
```

Comment:
As we're revisiting this, maybe we should comment on what exactly is unsafe about these functions. Namely, that the argument bytestring and the resulting raw bytes share the same internal structure, and therefore changes to one affect the other. The user shouldn't use the `ByteString` anymore after using this function

Reply:
But given that both are immutable types then this is safe right?